src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/GCHandleTests.cs

```diff
@@ -0,0 +1,191 @@
+            yield return new object[] { new NonBlittable() };
+            yield return new object[] { new object[0] };
+            yield return new object[] { new NonBlittable[0] };
+        }
+
+        [Theory]
+        [MemberData(nameof(InvalidPinnedObject_TestData))]
+        public void Alloc_InvalidPinnedObject_ThrowsArgumentException(object value)
+        {
+            AssertExtensions.Throws<ArgumentException>(null, () => GCHandle.Alloc(value, GCHandleType.Pinned));
+        }
+
+        [Theory]
+        [InlineData(GCHandleType.Weak - 1)]
+        [InlineData(GCHandleType.Pinned + 1)]
+        public void Alloc_InvalidGCHandleType_ThrowsArgumentOutOfRangeException(GCHandleType type)
+        {
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("type", () => GCHandle.Alloc(new object(), type));
+        }
+
+        [Fact]
+        public void FromIntPtr_Zero_ThrowsInvalidOperationException()
+        {
+            Assert.Throws<InvalidOperationException>(() => GCHandle.FromIntPtr(IntPtr.Zero));
+        }
+
+        [Fact]
+        public void AddrOfPinnedObject_NotInitialized_ThrowsInvalidOperationException()
+        {
+            var handle = new GCHandle();
+            Assert.Throws<InvalidOperationException>(() => handle.AddrOfPinnedObject());
+        }
+
+        [Fact]
+        public void AddrOfPinnedObject_NotPinned_ThrowsInvalidOperationException()
+        {
+            GCHandle handle = GCHandle.Alloc(new object());
+            Assert.Throws<InvalidOperationException>(() => handle.AddrOfPinnedObject());
+        }
+
+        [Fact]
+        public void Free_NotInitialized_ThrowsInvalidOperationException()
+        {
+            var handle = new GCHandle();
+            Assert.Throws<InvalidOperationException>(() => handle.Free());
+        }
+
+        public static IEnumerable<object[]> Equals_TestData()
+        {
+            GCHandle handle = GCHandle.Alloc(new object());
+            yield return new object[] { handle, handle, true };
+            yield return new object[] { GCHandle.Alloc(new object()), GCHandle.Alloc(new object()), false };
+
+            yield return new object[] { GCHandle.Alloc(new object()), new object(), false };
+            yield return new object[] { GCHandle.Alloc(new object()), null, false };
+        }
+
+        [Theory]
+        [MemberData(nameof(Equals_TestData))]
+        public void Equals_Object_ReturnsExpected(GCHandle handle, object other, bool expected)
+        {
+            try
+            {
+                Assert.Equal(expected, handle.Equals(other));
+                if (other is GCHandle otherHandle)
+                {
+                    Assert.Equal(expected, handle == otherHandle);
+                    Assert.Equal(!expected, handle != otherHandle);
+                }
+            }
+            finally
+            {
+                handle.Free();
+                if (other is GCHandle otherHandle && !ReferenceEquals(handle, otherHandle))
```

Comment:
GCHandle is a struct.  `ReferenceEquals(handle, otherHandle)` will always return false, as the two structs will be boxed into different objects.